subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java

```diff
@@ -866,58 +866,53 @@ public boolean add(Object action) {
             if (action == null) {
                 throw new InvalidUserDataException("Action must not be null!");
             }
-            return super.add(wrap((Action<? super Task>) action));
+            return super.add(wrap(Cast.uncheckedNonnullCast(action)));
         }
 
         @Override
         public void add(int index, Object action) {
             if (action == null) {
                 throw new InvalidUserDataException("Action must not be null!");
             }
-            super.add(index, wrap((Action<? super Task>) action));
+            super.add(index, wrap(Cast.uncheckedNonnullCast(action)));
         }
 
         @Override
         public boolean addAll(Collection actions) {
             if (actions == null) {
                 throw new InvalidUserDataException("Actions must not be null!");
             }
-            return super.addAll(transformToContextAwareTaskActions(actions));
+            return super.addAll(transformToContextAwareTaskActions(Cast.uncheckedNonnullCast(actions)));
```

Comment:
I'd love to eventually revert the nullability of this method, i.e. make `Cast.uncheckedCast()` be the one returning `@Nonnull`, and have a method with a longer name like `Cast.uncheckedNullableCast()` for the `@Nullable` case.

Reply:
Sounds 'nicer'. But I don't have a strong opinion. But we can do that in one refactor PR.

Reply:
What I mean is, we should use the `Cast.uncheckedNullableCast()` in most cases, so that one should have the simpler name. Asking for a `@Nullable` cast should be the exception.

Reply:
Not something we need to fix here, just thought I'd inject the idea as I see you are making similar sweeping changes here.

Reply:
Understood and agreed. So we should check all usages of `uncheckedCast()` if they should rather be `uncheckedNullableCast()`.

Reply:
Yep, that'd be great. Again, something to do in a separate PR.